### PR TITLE
core: remove CGroups by depth

### PR DESCRIPTION
**Pull request by emilengler**
This commit provides a partial fix for #17, by fixing the errors that happened during the removal of a CGroup.

The fix works by obtaining the sub-directories of a CGroup in a sorted manner, meaning that the outer-most directories are being removed before the lower ones are. This step is necessary, because removing non-empty directories is prohibited.

Beside this, it also adds a recovery for `EBUSY` errors, as they are not critical and simply indicate that a retry could be successful.

### CODE REVIEW

**Comment by dadada** [comment]
Wasn't #17 because the directory is created twice? Getting rid of the EBUSY is useful anyway, so :+1: from me.

File: core/src/cgroup.rs [hunk 1]
```diff
@@ -185,16 +188,30 @@ impl CGroup {
...
+            .flatten()
+            .filter(|e| e.file_type().is_dir())
+            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
+        {
+            trace!("Removing cgroup {}", &d.path().display());
+
+            // Re-try on every EBUSY
+            loop {
+                match fs::remove_dir(d.path()) {
+                    Err(e) => {
```
Comment (dadada) [inline core/src/cgroup.rs:206]: I guess the real question is why is the directory still busy, and how can we fix that.

File: core/src/cgroup.rs [hunk 2]
```diff
@@ -185,16 +188,30 @@ impl CGroup {
...
-            fs::remove_dir(dir)?;
+        // Delete all cgroups from deepest to highest.
+        // It is necessary to delete the outer-most directories first, because
+        // a non-empty directory may not be deleted.
+        trace!("Calling remove on {}", &self.path.display());
+        for d in WalkDir::new(&self.path)
+            .into_iter()
+            .flatten()
+            .filter(|e| e.file_type().is_dir())
+            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
```
Comment (dadada) [inline core/src/cgroup.rs:199]: :+1:

**Review by dadada — APPROVED** [review]

File: core/src/cgroup.rs [hunk 3]
```diff
@@ -185,16 +188,30 @@ impl CGroup {
...
+        {
+            trace!("Removing cgroup {}", &d.path().display());
+
+            // Re-try on every EBUSY
+            loop {
```
Comment (dadada) [inline core/src/cgroup.rs:204]: Is there maybe a child process we could wait for instead of busy waiting?

**Review by dadada — CHANGES_REQUESTED** [review]

File: core/src/cgroup.rs (hunk 1 above)
Comment (emilengler) [reply core/src/cgroup.rs:206]: I agree! The most weird thing remains however, that this does not occur (for my machine) when running the hypervisor with super user privileges...

File: core/src/cgroup.rs (hunk 3 above)
Comment (emilengler) [reply core/src/cgroup.rs:204]: It could be possible figuring it out using `fuser`.

Comment (dadada) [reply core/src/cgroup.rs:204]: We should be able to tell which processes we started and which resources we assigned to them. Therefore, we should already know which process we have to wait for without using `fuser`.

File: core/src/cgroup.rs (hunk 1 above)
Comment (dadada) [reply core/src/cgroup.rs:206]: Kind of the same question as https://github.com/aeronautical-informatics/apex-linux/pull/22#discussion_r1058656617

File: core/src/cgroup.rs (hunk 3 above)
Comment (emilengler) [reply core/src/cgroup.rs:204]: > We should be able to tell which processes we started and which resources we assigned to them. Therefore, we should already know which process we have to wait for without using fuser.

I'm kinda curious about that. As far as I am aware of, there is no POSIX way of obtaining a list containing all alive child processes, meaning we would have to manage such a list ourselves, which is a very error-prone task. I'll take a look on how fuser works under the hood.

Comment (emilengler) [reply core/src/cgroup.rs:204]: `/proc/<PID>/task/<PID>/children` looks very promising.

**Comment by emilengler** [comment]
@dadada Busy loop replaced with a poll loop.

**Comment by emilengler** [comment]
I would really enjoy having that PR merged ASAP @wucke13 @dadada.

**Comment by emilengler** [comment]
@dadada I suggest removing the poller (again) with a busy loop, because a poller is simply wrong for the following reasons:

First and foremost, a poller only checks if data is readable on a file descriptor. This condition will always be true for regular files on the filesystem.

Because of this, a poller is not the correct approach. Instead, we need something that notifies the interfacing application about changes in the filesystem. More concrete, an API that returns whenever the `cgroup.procs` file gets updated. The first (and only) solution that came to my mind, was to use `inotify` for this. However, it turns out, that inotify only works on true filesystems and not virtual filesystems like `/proc` and `/sys`.[^1]

[^1]: https://stackoverflow.com/questions/35878292/get-notification-on-cgroup-process-change

**Comment by emilengler** [comment]
51c4c43590c1449de5a2c7181623c4e44af9c07c contains a detailed commit message about it.

**Comment by dadada** [comment]
> First and foremost, a poller only checks if data is readable on a file descriptor. This condition will always be true for regular files on the filesystem.
> 

Yes, you are right that we should check for file modifications instead and reading from the file does only block if the device backing the file would block, but not if it is empty. As an alternative to cgroup.procs, maybe we should consider using `cgroup.events`, for which a file-modified event will be generated when the cgroup is no longer populated.

https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v2.rst

**Comment by emilengler** [comment]
https://github.com/aeronautical-informatics/apex-linux/blob/51c4c43590c1449de5a2c7181623c4e44af9c07c/core/src/cgroup.rs#L105

**Comment by emilengler** [comment]
squashed

**Comment by emilengler** [comment]
Ready for merge now. The `EBUSY` was triggered, because there were still living processes within the cgroup.

File: core/src/cgroup.rs [hunk 4]
```diff
@@ -99,7 +101,7 @@ impl CGroup {
             bail!("{} is not a valid cgroup", self.path.display());
         }
 
-        Ok(!fs::read(self.path.join("cgroup.procs"))?.is_empty())
+        Ok(fs::read_to_string(self.get_events_path())?.contains("populated 1\n"))
```
Comment (dadada) [inline core/src/cgroup.rs:104]: Wrong for `populated 2`.

File: core/src/cgroup.rs [hunk 5]
```diff
@@ -150,20 +152,34 @@ impl CGroup {
...
-        Ok(fs::write(path, "1")?)
+        // Emit the kill signal to all processes inside the cgroup
+        trace!("writing '1' to {}", killfile.display());
+        fs::write(killfile, "1")?;
+
+        // Check if all processes were terminated successfully
+        let mut i = 1;
+        while i <= 64 {
+            trace!("kill iteration ({i}/64)");
+            if !self.populated()? {
```
Comment (dadada) [inline core/src/cgroup.rs:177]: This returns too early, if populated > 1.

File: core/src/cgroup.rs (hunk 4 above)
Comment (emilengler) [reply core/src/cgroup.rs:104]: No. It is boolean.

**Review by dadada — APPROVED** [review]

**Pull request merged**